core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1470,21 +1503,45 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new InitPidResponse(throttleTimeMs, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED))
+    else {
+      val internalTopics = partitionsToAdd.asScala.filter {tp => kafka.common.Topic.InternalTopics.contains(tp.topic())}
+
+      val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) =
```

Comment:
So interestingly, we currently only have a single error code in the response schema, which means we can't return topic-level errors. Maybe we need to change this? As far as the request semantics go, I think it's reasonable to say that either all partitions are successfully added to the transaction or none of them are. In that sense, a single error code is fine, but it seems unfortunate if we can't tell the user which topics actually had the errors. What do you think?